Yubico.YubiKey/docs/users-manual/sdk-programming-guide/pin-complexity-policy.md

```diff
@@ -16,42 +16,60 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 
-### FIDO2 Session
-In the FIDO2 application, `Fido2Exception` with `Status` of `CtapStatus.PinPolicyViolation` is thrown after a PIN complexity was violated. For `KeyCollectors`, `KeyEntryData.IsViolatingPinComplexity` will be set to `true` for these situations.
+### Handling PIN complexity errors
 
-This applies to following `Fido2Session` operations:
-- `Fido2Session.SetPin()`
-- `Fido2Session.ChangePin()`
+Applications that support setting or changing PINs should be able to handle the situation when a YubiKey refuses the user value because it violates the PIN complexity policy.
 
-## Example code
-You can find examples of code in the `PivSampleCode` and `Fido2SampleCode` examples as well in `PinComplexityTests` integration tests.
\ No newline at end of file
+The SDK communicates PIN complexity violations by throwing specific exceptions.
+
+#### PivSession exceptions
+
+During a [PivSession](xref:Yubico.YubiKey.Piv.PivSession), PIN complexity violations result in a `System.Security.SecurityException` with the message, `ExceptionMessages.PinComplexityViolation`.
+
+If the application uses a [KeyCollector](xref:UsersManualKeyCollector), the violation is reported through the [KeyEntryData.IsViolatingPinComplexity](xref:Yubico.YubiKey.KeyEntryData.IsViolatingPinComplexity) property.
+
+PIN complexity violations are reported for following PIV operations:
+
+- [PivSession.ChangePin()](xref:Yubico.YubiKey.Piv.PivSession.ChangePin)
+- [PivSession.ChangePuk()](xref:Yubico.YubiKey.Piv.PivSession.ChangePuk)
+- [PivSession.ResetPin()](xref:Yubico.YubiKey.Piv.PivSession.ResetPin)
+
+#### Fido2Session exceptions
+
+During a [Fido2Session](xref:Yubico.YubiKey.Fido2.Fido2Session), PIN complexity violations result in a [Fido2Exception](xref:Yubico.YubiKey.Fido2.Fido2Exception) object with a [Status](xref:Yubico.YubiKey.Fido2.Fido2Exception.Status) of [CtapStatus.PinPolicyViolation](xref:Yubico.YubiKey.Fido2.CtapStatus.PinPolicyViolation).
+
+If the application uses a [KeyCollector](xref:UsersManualKeyCollector), the violation is reported through the [KeyEntryData.IsViolatingPinComplexity](xref:Yubico.YubiKey.KeyEntryData.IsViolatingPinComplexity) property.
+
+PIN complexity violations are reported for following FIDO2 operations:
+
+- [Fido2Session.SetPin()](xref:Yubico.YubiKey.Fido2.Fido2Session.SetPin)
+- [Fido2Session.ChangePin()](xref:Yubico.YubiKey.Fido2.Fido2Session.ChangePin)
+
+### Example code
+
+For code samples demonstrating how to handle PIN complexity violations, see the [PivSampleCode](https://github.com/Yubico/Yubico.NET.SDK/blob/main/Yubico.YubiKey/examples/PivSampleCode/KeyCollector/SampleKeyCollector.cs), [Fido2SampleCode](https://github.com/Yubico/Yubico.NET.SDK/blob/main/Yubico.YubiKey/examples/Fido2SampleCode/KeyCollector/Fido2SampleKeyCollector.cs), and [PinComplexityTests](https://github.com/Yubico/Yubico.NET.SDK/blob/main/Yubico.YubiKey/tests/integration/Yubico/YubiKey/PinComplexityTests.cs) integration tests.
```

Comment:
I linked to the files on the main branch instead of develop--let me know if that's the correct branch to use.

Reply:
The main branch is correct 👍🏻  I like how this chapter is now organized, thank you!

Reply:
Yes, very good!